feat(conductor): fix leader loop switch by adding round-robin leadership transfer

KyrinCode opened the pull request:
**Description**

This PR adds an optional round-robin leader transfer mechanism to op-conductor, configurable via the `--raft.round-robin-leader-transfer` flag.

**Problem**

The current leader transfer in op-conductor relies on Raft's default `LeadershipTransfer()`, which selects the next leader based on log completeness (the follower with the most up-to-date log). However, this selection mechanism is unaware of application-level health status (op-node & execution layer).
This can cause a problematic scenario where unhealthy nodes keep being elected as leader in a loop:
1. Node A becomes leader but is unhealthy → transfers to Node B (most complete log)
2. Node B becomes leader but is also unhealthy → transfers back to Node A (now has most complete log)
3. Healthy Node C never gets elected because its log is slightly behind

**Solution**

Introduce a deterministic round-robin leader transfer that cycles through all voters in sorted order (by ServerID):
- Node A → Node B → Node C → Node A → ...
This ensures that even if only one node in the cluster is healthy, it will eventually become the leader after at most N-1 transfers.

**Changes**

- Added `--raft.round-robin-leader-transfer` flag (default: false, backward compatible)
- Added transferLeaderRoundRobin() function that:
  - Sorts all voters by ServerID for consistent ordering across nodes
  - Transfers to the next node in sorted order
  - If transfer fails (e.g., target has stale logs), tries the next node
  - Handles "leadership transfer in progress" gracefully

**Tests**

Manual testing was performed with a 4-node conductor cluster where 2 nodes were intentionally made unhealthy. With round-robin enabled, leadership eventually transferred to a healthy node. Without round-robin, leadership kept bouncing between the unhealthy nodes.

**Additional context**

This feature is opt-in and disabled by default to maintain backward compatibility. When disabled, the behavior is identical to the current implementation.
The round-robin approach trades off Raft's "most up-to-date log" optimization for guaranteed leader rotation. In practice, this is acceptable because:
All voters in a healthy cluster should have similar log states
The primary goal of leader transfer in op-conductor is to find a healthy sequencer, not to optimize for log completeness
If a target node's log is too stale, Raft will reject the transfer and we try the next node

**Metadata**

- Related to sequencer high-availability and failover reliability

KyrinCode committed:
fix leader loop switch by adding round-robin leadership transfer

KyrinCode committed:
update

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

KyrinCode commented:
Hi team, could you take some time to review this? 🥲

yashvardhan-kukreja commented:
/ci authorize 79484995210518cc06e802915d964bac9bba5666

yashvardhan-kukreja commented:
Hi folks beginning the review

yashvardhan-kukreja commented on op-conductor/conductor/service.go:
```diff
@@ -837,6 +844,95 @@ func (oc *OpConductor) transferLeader() error {
+			allVoters = append(allVoters, server)
+		}
+	}
+
+	if len(allVoters) <= 1 {
+		return errors.New("no other voters available for leader transfer")
+	}
```
should it really exit with an error? I think it should exit with a no-op / warning, given today's behaviour.

yashvardhan-kukreja commented on op-conductor/conductor/service.go:
```diff
@@ -837,6 +844,95 @@ func (oc *OpConductor) transferLeader() error {
+		)
+	}
+
+	return errors.New("failed to transfer leadership to any voter")
```
In this case, what do you think about falling back to `oc.cons.TransferLeader()` instead of raising an error?

yashvardhan-kukreja reviewed (commented):
> Without round-robin, leadership kept bouncing between the unhealthy nodes

Interesting -- was this consistently reproducible?

-----

One concern worth calling out: this flag would need to be enabled on every conductor in the cluster to work as intended. A mixed configuration can actually make things worse.

Consider a cluster where conductor1 has round-robin enabled, but conductor2 is running an older version (or simply has the flag disabled):

- Conductor1 (round-robin) becomes leader, finds itself unhealthy, and deterministically transfers leadership to conductor2 (next in sorted order).
- Conductor2 (default behavior) becomes leader, is also unhealthy, and transfers leadership back to conductor1 -- since conductor1 now has the most up-to-date log after having just been a very recent leader.

This creates a leadership-transfer hot loop between conductor1 and conductor2, even if conductor3 is perfectly healthy.
This is arguably worse than the original problem, because:
- The bounce is now guaranteed rather than circumstantial -- round-robin will always pick conductor2, and default selection will always pick conductor1 (most up-to-date log).
- Log outdatedness is itself a health check criterion (via unsafe head tracking in the health monitor, where unsafe head happens to be the raft log entry committed by the leader's op-node)
- By ignoring log completeness, conductor1 is effectively transferring leadership to a node that is more likely to fail its health check, reinforcing the loop.

To be clear -- a similar bounce can occur without round-robin feature today as well (as you saw in your tests). 

Hence, I'm not blocking this PR, just flagging this mixed-configuration scenario in case it wasn't considered. Otherwise, this PR looks good to me.

-------

One potential idea to mitigate this issue, at the risk of complicating things, is for op-conductor to track the last time it relinquished leadership with which mechanism (round-robin or up-to-date log). If it's been very recent, then op-conductor should shift the leadership again this time but via a different mechanism to break the hot loop potentially.

Feel free to let me know what do you think.

Thanks again for your contribution.

KyrinCode committed:
fix(conductor): address reviewer feedback on round-robin leader transfer

- Change no-voters error to warn + no-op (matches default Raft behavior)
- Fall back to default TransferLeader() when all round-robin attempts fail
- Document mixed-configuration risk in flag description

KyrinCode commented:
Thanks for the thorough review @yashvardhan-kukreja !

**On reproducibility**: Yes, this is consistently reproducible. The root cause is that Raft's `LeadershipTransfer()` selects candidates based on log completeness, without considering application-level health. When multiple nodes are unhealthy, the node with the most up-to-date log keeps getting elected — which is often the node that was just leader (and is still unhealthy), creating a deterministic bounce.

**Changes made based on your feedback**:

1. **Rebased** onto the latest `develop`.
2. **`len(allVoters) <= 1`**: Changed from error to `Warn` + no-op (`return nil`), matching the default Raft behavior — no point retrying when there's nobody to transfer to.
3. **All round-robin attempts fail**: Now falls back to `oc.cons.TransferLeader()` (default Raft log-based selection) instead of returning an error.

**On the mixed-configuration concern**: Your analysis of the hot-loop scenario is spot on. That said, I'd prefer not to introduce transfer-history tracking to break the loop — it adds state management complexity that's hard to justify given the typical deployment: conductor clusters are small (3–5 nodes) and managed by a single ops team, so flag consistency is an operational baseline. Instead, I've added a note to both the flag comment and `Usage` string making it clear this must be enabled on **all** conductors in the cluster. Happy to revisit if this turns out to be insufficient in practice.

yashvardhan-kukreja commented:
Thanks @KyrinCode , sounds good.

yashvardhan-kukreja commented:
/ci authorize 29c71109937740c007927903873747957d20952b
